File: 1.CPU_Memory_Disk_Network/Network/Network_Programming_In_Practice/NPIP_Java/src/main/java/test/t02_tcp_data_transfer/TcpDataTransfer1Server.java
package test.t02_tcp_data_transfer;

import static java.lang.System.out;

import java.io.*;
import java.net.ServerSocket;
import java.net.Socket;

/**
 * TCP data transfer server. This server accept only one client connection intentionally.
 *
 * @author Camio1945
 */
public class TcpDataTransfer1Server {
  protected static final int PORT = 9123;

  public static void main(String[] args) throws Exception {
    try (ServerSocket serverSocket = new ServerSocket(PORT)) {
      out.printf("\nServer is listening on port %d ...\n", PORT);
      Socket acceptedSocket = serverSocket.accept();
      out.println("A client connected.");
      try (BufferedReader bufferedReader =
              new BufferedReader(new InputStreamReader(acceptedSocket.getInputStream()));
          PrintWriter printWriter = new PrintWriter(acceptedSocket.getOutputStream(), true)) {
        out.println("Received from client: " + bufferedReader.readLine());
        printWriter.println("Hello client, I am server.");
      }
    }
  }
}

File: 1.CPU_Memory_Disk_Network/Network/Network_Programming_In_Practice/NPIP_Go/t02_tcp_data_transfer/server/TcpDataTransfer1Server.go
package main

import (
	"fmt"
	"net"
)

const HOST = "localhost"
const PORT = 9123

// This server accept only one client connection intentionally.
func main() {
	listener, err := net.Listen("tcp", fmt.Sprintf("%s:%d", HOST, PORT))
	defer listener.Close()
	panicIfError(err, "Error listening on port", PORT)
	fmt.Printf("\nServer is listening on port %d ...\n", PORT)

	conn, err := listener.Accept()
	defer conn.Close()
	panicIfError(err, "Error accepting connection")
	fmt.Println("A client connected.")

	buffer := make([]byte, 1024)
	n, err := conn.Read(buffer)
	panicIfError(err, "Error reading from client")
	fmt.Printf("Received from client: %s\n", string(buffer[:n]))

	_, err = conn.Write([]byte("Hello client, I am server.\n"))
	panicIfError(err, "Write data to client failed")
}

func panicIfError(err error, messages ...any) {
	if err != nil {
		panic(fmt.Sprint(messages, "\nerror: ", err))
	}
}
